README.md
# log-util

[![NPM Version][npm-version-image]][npm-url]
[![NPM Downloads][npm-downloads-image]][npm-url]
[![MIT License][license-image]][license-url]

[![NPM][nodei-image]][nodei-url]

[![NPM][nodei-dl-image]][nodei-url]

Node.js log utility for terminal output.

![log-util](doc/log-util.png)

## usage

`yarn add log-util`

```
var log = require('log-util');
log.debug('debug', 0);
log.info('info', 1);
log.success('success', 2);
log.warn('warn', 3);
log.error('error', 4);
```

## api

### `log.setLevel(level: string) => log`

`level` is level id or level string

### `new log.Log(0) => log`

## log levels

```
log.levels.debug === 0;
log.levels.info === 1;
log.levels.success === 2;
log.levels.warn === 3;
log.levels.error === 4;
```

[npm-version-image]: http://img.shields.io/npm/v/log-util.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/log-util
[npm-downloads-image]: http://img.shields.io/npm/dm/log-util.svg?style=flat-square
[license-image]: http://img.shields.io/badge/license-MIT-blue.svg?style=flat-square
[license-url]: LICENSE
[nodei-dl-image]: https://nodei.co/npm-dl/log-util.png?height=3
[nodei-url]: https://nodei.co/npm/log-util/
[nodei-image]: https://nodei.co/npm/log-util.svg?downloads=true&downloadRank=true&stars=true


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.2.1](https://github.com/vivaxy/log-util/compare/v2.2.0...v2.2.1) (2019-06-26)


### Bug Fixes

* **level:** :bug: Update default level ([6c67526](https://github.com/vivaxy/log-util/commit/6c67526))


### Tests

* **test:** :white_check_mark: Fix debug level ([f59c4c6](https://github.com/vivaxy/log-util/commit/f59c4c6))



## [2.2.0](https://github.com/vivaxy/log-util/compare/v2.1.0...v2.2.0) (2019-06-26)


### Features

* **levels:** :sparkles: Add debug level ([5da01a7](https://github.com/vivaxy/log-util/commit/5da01a7))



## [2.1.0](https://github.com/vivaxy/log-util/compare/v2.0.0...v2.1.0) (2019-06-26)


### Features

* **types:** :label: Add types ([29e79b1](https://github.com/vivaxy/log-util/commit/29e79b1))



## [2.0.0](https://github.com/vivaxy/log-util/compare/v1.2.0...v2.0.0) (2019-06-26)


### Features

* **core:** :recycle: Completely rewrite ([cfd8e8c](https://github.com/vivaxy/log-util/commit/cfd8e8c))


### BREAKING CHANGES

* **core:** All changed.



<a name="1.2.0"></a>
# [1.2.0](https://github.com/vivaxy/log-util/compare/v1.1.1...v1.2.0) (2017-05-04)


### Features

* **scripts:** :sparkles:Add release script, using `standard-version` to bump version automatically ([29cd023](https://github.com/vivaxy/log-util/commit/29cd023))


test/index.js
/**
 * @since 15-08-19 16:37
 * @author vivaxy
 */
'use strict';
var log = require('../index.js');

console.log('--- test output ---');
log.verbose('verbose', log.Log.VERBOSE);
log.debug('debug', log.Log.DEBUG);
log.info('info', log.Log.INFO);
log.warn('warn', log.Log.WARN);
log.error('error', log.Log.ERROR);

console.log('--- test set level and set date format ---');
log.setLevel(3);
log.setDateFormat('yyyy-mm-dd');
log.verbose('verbose', log.Log.VERBOSE);
log.debug('debug', log.Log.DEBUG);
log.info('info', log.Log.INFO);
log.warn('warn', log.Log.WARN);
log.error('error', log.Log.ERROR);

console.log('--- test new instance ---');
var anotherLog = new log.Log(1);
anotherLog.verbose('verbose', log.Log.VERBOSE);
anotherLog.debug('debug', log.Log.DEBUG);
anotherLog.info('info', log.Log.INFO);
anotherLog.warn('warn', log.Log.WARN);
anotherLog.error('error', log.Log.ERROR);

console.log('--- new instance should not affects others ---');
log.verbose('verbose', log.Log.VERBOSE);
log.debug('debug', log.Log.DEBUG);
log.info('info', log.Log.INFO);
log.warn('warn', log.Log.WARN);
log.error('error', log.Log.ERROR);

console.log('--- access to level array, new instances should not have levelArray ---');
console.log(log.levelArray);
console.log(anotherLog.levelArray);

console.log('--- output objects ---');
log.verbose({
    string: 'string',
    number: 1,
    boolean: true,
    functions: function () {
        log.debug('output function');
    },
    date: new Date(),
    array: [1, 'string', false],
    object: {
        string: 'string',
        number: 2
    }
});


src/__tests__/index.ts
/**
 * @since 15-08-19 16:37
 * @author vivaxy
 */

import test from 'ava';
import chalk from 'chalk';
import * as figures from 'figures';
import * as logSymbol from 'log-symbols';
import * as log from '../index';

let __memo__: any[] = [];
let globalConsoleLog = global.console.log;

test.beforeEach(() => {
  global.console.log = function mockConsoleLog(...messages: any[]) {
    __memo__ = __memo__.concat(messages);
  };
});

test.afterEach(() => {
  while (__memo__.length) {
    __memo__.pop();
  }
  global.console.log = globalConsoleLog;
});

test('should output debug levels', (t) => {
  log.setLevel(log.levels.debug);
  log.debug('debug', log.levels.debug);
  t.deepEqual(__memo__, [
    chalk.grey(figures.questionMarkPrefix),
    'debug',
    log.levels.debug,
  ]);
});

test('should output info levels', (t) => {
  log.info('info', log.levels.info);
  t.deepEqual(__memo__, [logSymbol.info, 'info', log.levels.info]);
});

test('should output success levels', (t) => {
  log.success('success', log.levels.success);
  t.deepEqual(__memo__, [logSymbol.success, 'success', log.levels.success]);
});

test('should output warn levels', (t) => {
  log.warn('warn', log.levels.warn);
  t.deepEqual(__memo__, [logSymbol.warning, 'warn', log.levels.warn]);
});

test('should output error levels', (t) => {
  log.error('error', log.levels.error);
  t.deepEqual(__memo__, [logSymbol.error, 'error', log.levels.error]);
});

test('should follow log level info', (t) => {
  log.setLevel(log.levels.info);
  log.debug();
  log.info();
  log.success();
  log.warn();
  log.error();
  t.deepEqual(__memo__, [
    logSymbol.info,
    logSymbol.success,
    logSymbol.warning,
    logSymbol.error,
  ]);
});

test('should follow log level success', (t) => {
  log.setLevel(log.levels.success);
  log.debug();
  log.info();
  log.success();
  log.warn();
  log.error();
  t.deepEqual(__memo__, [
    logSymbol.success,
    logSymbol.warning,
    logSymbol.error,
  ]);
});

test('should follow log level warn', (t) => {
  log.setLevel(log.levels.warn);
  log.debug();
  log.info();
  log.success();
  log.warn();
  log.error();
  t.deepEqual(__memo__, [logSymbol.warning, logSymbol.error]);
});

test('should follow log level error', (t) => {
  log.setLevel(log.levels.error);
  log.debug();
  log.info();
  log.success();
  log.warn();
  log.error();
  t.deepEqual(__memo__, [logSymbol.error]);
});


index.js
/**
 * @since 15-08-19 16:13
 * @author vivaxy
 */

'use strict';

var util = require('util');
var chalk = require('chalk');
var dateFormat = require('dateformat');

/**
 * level object array
 * @type {*[]}
 */
var levelArray = [
    {
        level: 0,
        string: 'verbose',
        color: 'gray'
    },
    {
        level: 1,
        string: 'debug',
        color: 'green'
    },
    {
        level: 2,
        string: 'info',
        color: 'cyan'
    },
    {
        level: 3,
        string: 'warn',
        color: 'yellow'
    },
    {
        level: 4,
        string: 'error',
        color: 'red'
    }
];

/**
 * Log
 * @param level
 * @param dateFormat
 * @constructor
 */
var Log = function Log (level, dateFormat) {
    this.setLevel(level);
    this.setDateFormat(dateFormat);
};

Log.prototype.constructor = Log;

/**
 * main log method
 * @param level
 * @returns {Log}
 */
Log.prototype.log = function (level) {
    var time = chalk.gray('[' + dateFormat(new Date(), this.dateFormat) + ']');
    var args = Array.prototype.slice.call(arguments, 1);
    var color = this.find('level', level).color;
    args = args.map(function (arg) {
        if (typeof arg === 'object') {
            arg = util.inspect(arg, {
                depth: null
            });
            console.log(arg);
        }
        return chalk[color](arg);
    });
    args.unshift(time);
    if (level >= this.level) {
        console.log.apply(console, args);
    }
    return this;
};

/**
 * find level object which `key` = `value`
 * @param key
 * @param value
 * @returns {*}
 */
Log.prototype.find = function (key, value) {
    return levelArray.filter(function (o) {
            return o[key] === value;
        })[0] || {};
};

/**
 * set level
 * @param level
 * @returns {Log}
 */
Log.prototype.setLevel = function (level) {
    if (typeof level === 'string') level = this.find('string', level.toLowerCase()).level;
    this.level = level || 0;
    return this;
};

/**
 * set date format
 * @param dateFormat
 * @returns {Log}
 */
Log.prototype.setDateFormat = function (dateFormat) {
    this.dateFormat = dateFormat || 'HH:MM:ss.l';
    return this;
};

levelArray.forEach(function (o) {

    /**
     * export level string value
     * @type {number|*|Array}
     */
    Log[o.string.toUpperCase()] = o.level;

    /**
     * set methods to prototype
     * @returns {number|*}
     */
    Log.prototype[o.string] = function () {
        var args = arguments;
        Array.prototype.unshift.call(args, o.level);
        return this.log.apply(this, args);
    };
});

var log = new Log();
log.Log = Log;
log.levelArray = levelArray;
module.exports = log;


src/index.ts
/**
 * @since 15-08-19 16:13
 * @author vivaxy
 */

import chalk from 'chalk';
import * as util from 'util';
import * as figures from 'figures';
import * as logSymbols from 'log-symbols';

export enum levels {
  debug = 0,
  info = 1,
  success = 2,
  warn = 3,
  error = 4,
}

export let level: levels = 0;

type Symbol = keyof typeof logSymbols | 'debug';

function getIcon(symbol: Symbol): string {
  if (symbol === 'debug') {
    return chalk.grey(figures.pointerSmall);
  }
  return logSymbols[symbol];
}

function createLogger(_level: levels, symbol: Symbol) {
  return function(...messages: any[]): void {
    if (_level >= level) {
      const formatted = messages.map(function(message) {
        if (typeof message === 'object') {
          return util.inspect(message, {
            depth: null,
          });
        }
        return message;
      });
      console.log(getIcon(symbol), ...formatted);
    }
  };
}

export function setLevel(_level: levels): void {
  level = _level;
}

export const debug = createLogger(levels.debug, 'debug');
export const info = createLogger(levels.info, 'info');
export const success = createLogger(levels.success, 'success');
export const warn = createLogger(levels.warn, 'warning');
export const error = createLogger(levels.error, 'error');
